### DISCUSSION TITLE
Breakout dependency sets into extras, establish a better pattern for adapters with missing imports

### DISCUSSION THREAD

**Discussion by timkpaine** [Point72/csp#165] [Ideas]
Moved from https://github.com/Point72/csp/issues/35 to better facilitate back-and-forth.

We haven't yet decided whether we want to push for adapters to all be part of `csp` proper, or if we want/expect adapters to be maintained independently. Regardless, we currently have C++-based `parquet` and `kafka` adapters, python-based `db`, `csv`, `perspective`, `websocket`, `symphony`, and `slack` adapters, with plans for C++-based `redis`, and`websocket` adapters, so it seems likely that we will have a reliance on lots of python libraries.

We will want to establish a (better) pattern for optional dependency sets, at the very least moving optional dependencies to their own extras blocks in `pyproject.toml`, and having a more graceful handling at runtime when dependencies are missing.

**Comment by trallard** [comment]
This was discussed with @robambalu 
 and @ptomecek, we agreed:

- Will aim for a plugin-like approach for adapters vs having them all be part of `csp`
  - This will help with long-term maintenance (reducing the burden on `csp` maintainers, especially regarding dependencies compatibility)
  - This will also help with user experience, as otherwise, users would end up with a lot of dependencies and adapters that they do not want/need
- If there are (external) adapters that prove to be integral to `csp` and their users, then these could be added to `csp` (will need to define clear guidelines and document)
- To start with:
  - `parquet` and `kafka` should remain part of `csp`
  - the rest could be moved out; we can start with moving the `slack` and `symphony` adapters out of `csp`

**Reply by timkpaine** [reply]
Seems like clear guidelines should happen before moving things in/out?

From a dependency standpoint for pure python deps, `db` is far worse than the others, `perspective` is massively bit-rotted, `websocket` is very incomplete.

`slack` and `symphony` dependencies are optional *today*, and their tests are more complete than the others, so they actually have the **least** maintenance and dependency overhead.

**Comment by timkpaine** [comment]
WIP starting on a separate pure-python adapter
